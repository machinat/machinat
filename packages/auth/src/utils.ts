import type { ServerResponse, IncomingMessage } from 'http';
import { posix as posixPath } from 'path';
import getRawBody from 'raw-body';
import {
  parse as parseCookie,
  serialize as serializeCookie,
  CookieSerializeOptions,
} from 'cookie';
import type {
  WithHeaders,
  AuthApiResponseBody,
  AuthApiErrorBody,
} from './types.js';

export const respondRedirect = (res: ServerResponse, url: string): void => {
  res.writeHead(302, { Location: url });
  res.end();
};

export const getCookies = (req: WithHeaders): null | Record<string, string> => {
  if (!req.headers.cookie) {
    return null;
  }
  return parseCookie(req.headers.cookie);
};

const SET_COOKIE = 'Set-Cookie';

export const setCookie = (
  res: ServerResponse,
  key: string,
  value: string,
  options: CookieSerializeOptions
): void => {
  const cookieDesc = serializeCookie(key, value, options);
  const cookiesAlreadySet = res.getHeader(SET_COOKIE);

  res.setHeader(
    SET_COOKIE,
    Array.isArray(cookiesAlreadySet)
      ? [...cookiesAlreadySet, cookieDesc]
      : typeof cookiesAlreadySet === 'string'
      ? [cookiesAlreadySet, cookieDesc]
      : cookieDesc
  );
};

export const isSubdomain = (domain: string, subdomain: string): boolean =>
  subdomain.endsWith(domain) &&
  (subdomain.length === domain.length ||
    subdomain[subdomain.length - domain.length - 1] === '.');

export const isSubpath = (
  path: null | string,
  subpath: null | string
): boolean =>
  !!(path && subpath) && !posixPath.relative(path, subpath).startsWith('..');

const CONTENT_TYPE_JSON = { 'Content-Type': 'application/json' };

export const respondApiOk = (
  res: ServerResponse,
  platform: string,
  token: string
): void => {
  res.writeHead(200, CONTENT_TYPE_JSON);
  const body: AuthApiResponseBody = { platform, token };
  res.end(JSON.stringify(body));
};

export const respondApiError = (
  res: ServerResponse,
  platform: undefined | string,
  code: number,
  reason: string
): void => {
  res.writeHead(code, CONTENT_TYPE_JSON);
  const body: AuthApiErrorBody = { platform, error: { code, reason } };
  res.end(JSON.stringify(body));
};

export const parseJsonBody = async (req: IncomingMessage): Promise<any> => {
  try {
    const rawBody = await getRawBody(req, { encoding: true });
    const body = JSON.parse(rawBody);

    return typeof body === 'object' ? body : null;
  } catch (err) {
    return null;
  }
};
