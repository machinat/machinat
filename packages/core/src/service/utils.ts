import invariant from 'invariant';
import {
  SOCIABLY_SERVICE_CONTAINER,
  SOCIABLY_SERVICE_PROVIDER,
  SOCIABLY_SERVICE_INTERFACE,
} from '../symbol.js';
import ServiceScope from './scope.js';
import ServiceMaker from './maker.js';
import ProvisionMap from './provisionMap.js';
import type {
  Interfaceable,
  ServiceRequirement,
  ServiceContainer,
  ServiceProvider,
  MaybeContainer,
} from './types.js';

export const isServiceContainer = <T>(
  target: MaybeContainer<T>,
): target is ServiceContainer<T, unknown[]> =>
  typeof target === 'function' &&
  '$$typeof' in target &&
  target.$$typeof === SOCIABLY_SERVICE_CONTAINER;

export const isServiceProvider = (
  target: any,
): target is ServiceProvider<unknown, unknown[]> =>
  (typeof target === 'function' ||
    (typeof target === 'object' && target !== null)) &&
  target.$$typeof === SOCIABLY_SERVICE_PROVIDER;

export const maybeInjectContainer = <T>(
  scope: ServiceScope,
  maybeContainer: MaybeContainer<T>,
): T =>
  isServiceContainer(maybeContainer)
    ? scope.injectContainer(maybeContainer)
    : maybeContainer;

export const isInterfaceable = (target: any): target is Interfaceable<any> =>
  (typeof target === 'function' ||
    (typeof target === 'object' && target !== null)) &&
  (target.$$typeof === SOCIABLY_SERVICE_INTERFACE ||
    target.$$typeof === SOCIABLY_SERVICE_PROVIDER);

export const createEmptyScope = (): ServiceScope =>
  new ServiceScope(new ServiceMaker(new ProvisionMap()), new Map());

export const polishServiceRequirement = <T>(
  dep: Interfaceable<T> | ServiceRequirement<Interfaceable<T>>,
): ServiceRequirement<Interfaceable<T>> => {
  if (isInterfaceable(dep)) {
    return { require: dep, optional: false };
  }

  invariant(
    isInterfaceable(dep?.require),
    `${(dep as any).name || String(dep)} is not a valid interface`,
  );

  return dep;
};

export const validateLifetime = (lifetime: string): void => {
  invariant(
    lifetime === 'singleton' ||
      lifetime === 'scoped' ||
      lifetime === 'transient',
    `${lifetime} is not valid service lifetime`,
  );
};
