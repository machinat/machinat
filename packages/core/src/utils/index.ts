export { default as compose } from './compose.js';
export { default as formatNode } from './formatNode.js';
export { default as getTimeId } from './getTimeId.js';
export { default as mixin } from './mixin.js';
export * from './isX.js';
