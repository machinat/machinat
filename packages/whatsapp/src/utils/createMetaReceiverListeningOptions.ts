import { PopEventWrapper } from '@sociably/core';
import { ListeningPlatformOptions } from '@sociably/meta-api';
import WhatsAppBot from '../Bot.js';
import { WHATSAPP } from '../constant.js';
import eventFactory from '../event/factory.js';
import { WhatsAppEventContext } from '../types.js';

const createMetaReceiverListeningOptions = (
  bot: WhatsAppBot,
  popEventWrapper: PopEventWrapper<WhatsAppEventContext, null>
): ListeningPlatformOptions<WhatsAppEventContext> => ({
  platform: WHATSAPP,
  bot,
  objectType: 'whatsapp_business_account',
  makeEventsFromUpdate: eventFactory,
  popEvent: popEventWrapper(async () => null),
});

export default createMetaReceiverListeningOptions;
