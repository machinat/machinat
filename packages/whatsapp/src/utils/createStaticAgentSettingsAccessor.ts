import parsePhoneNumber from 'libphonenumber-js';
import { AgentSettingsAccessorI } from '../interface.js';
import {
  WhatsAppBusinessAccountSettings,
  WhatsAppAgentSettings,
} from '../types.js';

const normalizePhoneNumber = (
  settings: WhatsAppAgentSettings
): WhatsAppAgentSettings => {
  const parsedNumber = parsePhoneNumber(settings.phoneNumber);

  if (!parsedNumber) {
    throw new Error(`Invalid phone number format: ${settings.phoneNumber}`);
  }
  return {
    ...settings,
    phoneNumber: parsedNumber.number,
  };
};

export const singleStaticAgentSettingsAccessor = (
  settings: WhatsAppAgentSettings
): AgentSettingsAccessorI => {
  const normalizedSettings = normalizePhoneNumber(settings);

  return {
    getAgentSettings: async ({ numberId }) =>
      numberId === normalizedSettings.numberId ? normalizedSettings : null,
    getAgentSettingsBatch: async (numbers) =>
      numbers.map(({ numberId }) =>
        numberId === normalizedSettings.numberId ? normalizedSettings : null
      ),
  };
};

export const multiStaticAgentSettingsAccessor = (
  settings: WhatsAppBusinessAccountSettings[]
): AgentSettingsAccessorI => {
  const settingsMapping = new Map<string, WhatsAppAgentSettings>();
  for (const { accountId, numbers } of settings) {
    for (const { numberId, phoneNumber } of numbers) {
      settingsMapping.set(
        numberId,
        normalizePhoneNumber({
          accountId,
          numberId,
          phoneNumber,
        })
      );
    }
  }

  return {
    getAgentSettings: async ({ numberId }) =>
      settingsMapping.get(numberId) || null,
    getAgentSettingsBatch: async (numbers) =>
      numbers.map(({ numberId }) => settingsMapping.get(numberId) || null),
  };
};
