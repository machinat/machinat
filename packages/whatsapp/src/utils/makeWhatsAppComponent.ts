import { makeNativeComponent } from '@sociably/core/renderer';
import { WHATSAPP } from '../constant.js';

const makeWhatsAppComponent = makeNativeComponent(WHATSAPP);

export default makeWhatsAppComponent;
