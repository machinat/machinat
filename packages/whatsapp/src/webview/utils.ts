/* eslint-disable import/prefer-default-export */
import type { ContextDetails } from '@sociably/auth';
import WhatsAppChat from '../Chat';
import WhatsAppUser from '../User';
import WhatsAppAgent from '../Agent';
import type { WhatsAppAuthContext, WhatsAppAuthData } from './types';

export const getAuthContextDetails = (
  data: WhatsAppAuthData
): ContextDetails<WhatsAppAuthContext> => {
  const {
    account,
    agent: { id: agentNumberId, num: agentNumber },
    user: userNumberId,
  } = data;
  return {
    user: new WhatsAppUser(userNumberId),
    thread: new WhatsAppChat(agentNumberId, userNumberId),
    channel: new WhatsAppAgent(agentNumberId),
    agentNumber,
    businessAccountId: account,
  };
};

export const trimWaUrlNumber = (number: string): string => {
  if (number.startsWith('+')) {
    return number.slice(1);
  }
  return number;
};
