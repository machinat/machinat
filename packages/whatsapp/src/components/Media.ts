import { NativeElement, SociablyNode } from '@sociably/core';
import {
  makeUnitSegment,
  UnitSegment,
  InnerRenderFn,
} from '@sociably/core/renderer';
import makeWhatsAppComponent from '../utils/makeWhatsAppComponent';
import { WhatsAppSegmentValue, WhatsAppComponent } from '../types';

/**
 * @category Props
 */
export type MediaProps = {
  /** The uploaded media ID */
  mediaId?: string;
  /** The URL of the media to be sent */
  url?: string;
  /** The file content to be upload */
  fileData?: Buffer | NodeJS.ReadableStream;
  /** The content type of the file. Required when uploading using `file` */
  fileType?: string;
  /** Reply to the specified message */
  replyTo?: string;
  /** The asset tag for saving the created media and reusing it */
  assetTag?: string;
};

export type WithCaption = {
  /** Text to describes the media */
  caption?: SociablyNode;
};

export type WithFileName = {
  /**
   * Describes the filename for the specific document. The extension of the
   * filename will specify what format the document is displayed as in WhatsApp
   */
  fileName?: string;
};

const makeMediaComponent = (componentName: string, mediaType: string) => {
  const container = {
    [componentName]: async (
      node: NativeElement<MediaProps & WithCaption & WithFileName, any>,
      path: string,
      render: InnerRenderFn
    ) => {
      const {
        mediaId,
        url,
        fileData,
        fileType,
        caption,
        fileName,
        replyTo,
        assetTag,
      } = node.props;

      if (
        !(url || mediaId || fileData) ||
        (url && mediaId) ||
        (url && fileData) ||
        (fileData && mediaId)
      ) {
        throw new TypeError(
          'there should be exactly one of "url", "mediaId" or "fileData" prop'
        );
      }
      if (fileData && !fileType) {
        throw new TypeError(
          '"fileType" prop is required when using "fileData" source'
        );
      }

      const captionSegments = await render(caption, '.caption');
      if (captionSegments) {
        for (const seg of captionSegments) {
          if (seg.type !== 'text') {
            throw new TypeError(
              `"caption" prop should contain only textual content`
            );
          }
        }
      }

      return [
        makeUnitSegment(node, path, {
          message: {
            type: mediaType,
            [mediaType]: {
              id: mediaId,
              link: url,
              caption: captionSegments?.[0].value,
              filename: fileName,
            },
            context: replyTo ? { message_id: replyTo } : undefined,
          },
          mediaFile: fileData
            ? {
                data: fileData,
                type: fileType as string,
                info: { contentType: fileType, filename: fileName },
                assetTag,
              }
            : undefined,
        }),
      ];
    },
  };

  return makeWhatsAppComponent(container[componentName]);
};

/**
 * Send an audio
 * @category Component
 * @props {@link AudioProps}
 */
export const Audio: WhatsAppComponent<
  MediaProps,
  UnitSegment<WhatsAppSegmentValue>
> = makeMediaComponent('Audio', 'audio');

/**
 * Send a document
 * @category Component
 * @props {@link DocumentProps}
 */
export const Document: WhatsAppComponent<
  MediaProps & WithCaption & WithFileName,
  UnitSegment<WhatsAppSegmentValue>
> = makeMediaComponent('Document', 'document');

/**
 * Send an image
 * @category Component
 * @props {@link ImageProps}
 */
export const Image: WhatsAppComponent<
  MediaProps & WithCaption,
  UnitSegment<WhatsAppSegmentValue>
> = makeMediaComponent('Image', 'image');

/**
 * Send a sticker
 * @category Component
 * @props {@link StickerProps}
 */
export const Sticker: WhatsAppComponent<
  MediaProps,
  UnitSegment<WhatsAppSegmentValue>
> = makeMediaComponent('Sticker', 'sticker');

/**
 * Send a video
 * @category Component
 * @props {@link VideoProps}
 */
export const Video: WhatsAppComponent<
  MediaProps & WithCaption & WithFileName,
  UnitSegment<WhatsAppSegmentValue>
> = makeMediaComponent('Video', 'video');
