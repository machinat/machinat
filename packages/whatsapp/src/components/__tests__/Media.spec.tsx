import Sociably from '@sociably/core';
import { isNativeType } from '@sociably/core/utils';
import { Audio, Image, Document, Video, Sticker } from '../Media.js';
import { renderUnitElement } from './utils.js';

it('is a valid Component', () => {
  for (const Media of [Audio, Image, Document, Video, Sticker]) {
    expect(typeof Media).toBe('function');
    expect(isNativeType(<Media mediaId="123" />)).toBe(true);
    expect(Media.$$platform).toBe('whatsapp');
  }
});

test.each([
  ['audio', Audio],
  ['image', Image],
  ['document', Document],
  ['video', Video],
  ['sticker', Sticker],
])('rendering %s value', async (type, Media) => {
  await expect(renderUnitElement(<Media mediaId="123" />)).resolves.toEqual([
    {
      type: 'unit',
      node: <Media mediaId="123" />,
      path: '$',
      value: { message: { type, [type]: { id: '123' } } },
    },
  ]);
  await expect(
    renderUnitElement(<Media url="http://foo.bar/baz" replyTo="MESSAGE_ID" />)
  ).resolves.toEqual([
    {
      type: 'unit',
      node: <Media url="http://foo.bar/baz" replyTo="MESSAGE_ID" />,
      path: '$',
      value: {
        message: {
          type,
          [type]: { link: 'http://foo.bar/baz' },
          context: { message_id: 'MESSAGE_ID' },
        },
      },
    },
  ]);
  const fileData = Buffer.from('_BINARY_DATA_');
  await expect(
    renderUnitElement(<Media fileType="foo/bar" fileData={fileData} />)
  ).resolves.toEqual([
    {
      type: 'unit',
      node: <Media fileType="foo/bar" fileData={fileData} />,
      path: '$',
      value: {
        message: { type, [type]: {} },
        mediaFile: {
          type: 'foo/bar',
          data: fileData,
          info: { contentType: 'foo/bar' },
        },
      },
    },
  ]);
});

test.each([
  ['image', Image],
  ['document', Document],
  ['video', Video],
])('rendering %s value with caption', async (type, Media) => {
  let node = <Media mediaId="123" caption="FOO" />;
  await expect(renderUnitElement(node)).resolves.toEqual([
    {
      type: 'unit',
      node,
      path: '$',
      value: {
        message: { type, [type]: { id: '123', caption: 'FOO' } },
      },
    },
  ]);
  node = (
    <Media
      url="http://foo.bar/baz"
      caption={
        <>
          FOO {'BAR'} <>BAZ</>
        </>
      }
    />
  );
  await expect(renderUnitElement(node)).resolves.toEqual([
    {
      type: 'unit',
      node,
      path: '$',
      value: {
        message: {
          type,
          [type]: { link: 'http://foo.bar/baz', caption: 'FOO BAR BAZ' },
        },
      },
    },
  ]);
});

test('rendering document with fileName', async () => {
  await expect(
    renderUnitElement(<Document mediaId="123" fileName="foo.pdf" />)
  ).resolves.toEqual([
    {
      type: 'unit',
      node: <Document mediaId="123" fileName="foo.pdf" />,
      path: '$',
      value: {
        message: {
          type: 'document',
          document: { id: '123', filename: 'foo.pdf' },
        },
      },
    },
  ]);
});

test('throw when invalid props received', async () => {
  /* eslint-disable no-await-in-loop */
  for (const Media of [Audio, Image, Document, Video, Sticker]) {
    await expect(
      renderUnitElement(<Media />)
    ).rejects.toThrowErrorMatchingSnapshot();
    await expect(
      renderUnitElement(<Media mediaId="123" url="http://foo.bar/baz" />)
    ).rejects.toThrowErrorMatchingSnapshot();
    await expect(
      renderUnitElement(
        <Media fileData={Buffer.from('FOO')} url="http://foo.bar/baz" />
      )
    ).rejects.toThrowErrorMatchingSnapshot();
    await expect(
      renderUnitElement(
        <Media mediaId="123" fileData={Buffer.from('')} fileType="foo/bar" />
      )
    ).rejects.toThrowErrorMatchingSnapshot();
    await expect(
      renderUnitElement(<Media fileData={Buffer.from('FOO')} />)
    ).rejects.toThrowErrorMatchingSnapshot();
  }
  for (const Media of [Image, Document, Video]) {
    await expect(
      renderUnitElement(
        <Media mediaId="123" caption={<Image mediaId="456" />} />
      )
    ).rejects.toThrowErrorMatchingSnapshot();
  }
});
