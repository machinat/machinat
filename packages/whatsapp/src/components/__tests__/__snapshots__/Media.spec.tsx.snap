// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`throw when invalid props received 1`] = `"there should be exactly one of "url", "mediaId" or "fileData" prop"`;

exports[`throw when invalid props received 2`] = `"there should be exactly one of "url", "mediaId" or "fileData" prop"`;

exports[`throw when invalid props received 3`] = `"there should be exactly one of "url", "mediaId" or "fileData" prop"`;

exports[`throw when invalid props received 4`] = `"there should be exactly one of "url", "mediaId" or "fileData" prop"`;

exports[`throw when invalid props received 5`] = `""fileType" prop is required when using "fileData" source"`;

exports[`throw when invalid props received 6`] = `"there should be exactly one of "url", "mediaId" or "fileData" prop"`;

exports[`throw when invalid props received 7`] = `"there should be exactly one of "url", "mediaId" or "fileData" prop"`;

exports[`throw when invalid props received 8`] = `"there should be exactly one of "url", "mediaId" or "fileData" prop"`;

exports[`throw when invalid props received 9`] = `"there should be exactly one of "url", "mediaId" or "fileData" prop"`;

exports[`throw when invalid props received 10`] = `""fileType" prop is required when using "fileData" source"`;

exports[`throw when invalid props received 11`] = `"there should be exactly one of "url", "mediaId" or "fileData" prop"`;

exports[`throw when invalid props received 12`] = `"there should be exactly one of "url", "mediaId" or "fileData" prop"`;

exports[`throw when invalid props received 13`] = `"there should be exactly one of "url", "mediaId" or "fileData" prop"`;

exports[`throw when invalid props received 14`] = `"there should be exactly one of "url", "mediaId" or "fileData" prop"`;

exports[`throw when invalid props received 15`] = `""fileType" prop is required when using "fileData" source"`;

exports[`throw when invalid props received 16`] = `"there should be exactly one of "url", "mediaId" or "fileData" prop"`;

exports[`throw when invalid props received 17`] = `"there should be exactly one of "url", "mediaId" or "fileData" prop"`;

exports[`throw when invalid props received 18`] = `"there should be exactly one of "url", "mediaId" or "fileData" prop"`;

exports[`throw when invalid props received 19`] = `"there should be exactly one of "url", "mediaId" or "fileData" prop"`;

exports[`throw when invalid props received 20`] = `""fileType" prop is required when using "fileData" source"`;

exports[`throw when invalid props received 21`] = `"there should be exactly one of "url", "mediaId" or "fileData" prop"`;

exports[`throw when invalid props received 22`] = `"there should be exactly one of "url", "mediaId" or "fileData" prop"`;

exports[`throw when invalid props received 23`] = `"there should be exactly one of "url", "mediaId" or "fileData" prop"`;

exports[`throw when invalid props received 24`] = `"there should be exactly one of "url", "mediaId" or "fileData" prop"`;

exports[`throw when invalid props received 25`] = `""fileType" prop is required when using "fileData" source"`;

exports[`throw when invalid props received 26`] = `""caption" prop should contain only textual content"`;

exports[`throw when invalid props received 27`] = `""caption" prop should contain only textual content"`;

exports[`throw when invalid props received 28`] = `""caption" prop should contain only textual content"`;
